File: cql-ls-plugin-debug/src/main/java/org/opencds/cqf/cql/ls/plugin/debug/DebugCommandContribution.java
package org.opencds.cqf.cql.ls.plugin.debug;

import java.util.Collections;
import java.util.Set;
import java.util.concurrent.CompletableFuture;

import org.eclipse.lsp4j.ExecuteCommandParams;
import org.opencds.cqf.cql.ls.FuturesHelper;
import org.opencds.cqf.cql.ls.manager.CqlTranslationManager;
import org.opencds.cqf.cql.ls.plugin.CommandContribution;
import org.opencds.cqf.cql.ls.plugin.debug.session.DebugSession;

public class DebugCommandContribution implements CommandContribution {

    public static final String START_DEBUG_COMMAND = "org.opencds.cqf.cql.ls.plugin.debug.startDebugSession";

    private DebugSession debugSession = null;

    private CqlTranslationManager cqlTranslationManager;

    public DebugCommandContribution(CqlTranslationManager cqlTranslationManager) {
        this.cqlTranslationManager = cqlTranslationManager;
    }

    @Override
    public Set<String> getCommands() {
        return Collections.singleton(START_DEBUG_COMMAND);
    }

    @Override
    public CompletableFuture<Object> executeCommand(ExecuteCommandParams params) {

        this.cqlTranslationManager.hashCode();

        switch(params.getCommand()) {
            case START_DEBUG_COMMAND:
                if (this.debugSession == null || !this.debugSession.isActive()) {
                    this.debugSession = new DebugSession();
                    try {
                        return CompletableFuture.completedFuture(this.debugSession.start().get());
                    }
                    catch (Exception e) {
                        return FuturesHelper.failedFuture(e);
                    }
                }
                else {
                    throw new IllegalStateException(String.format("Please wait for the current debug session to end before starting a new one."));
                }
            default:
                throw new IllegalArgumentException(String.format("DebugPlugin doesn't support command %s", params.getCommand()));
        }
     }
}

File: ls/server/src/main/kotlin/org/opencds/cqf/cql/debug/DebugCommandContribution.kt
package org.opencds.cqf.cql.debug

import org.eclipse.lsp4j.ExecuteCommandParams
import org.opencds.cqf.cql.ls.core.ContentService
import org.opencds.cqf.cql.ls.server.manager.CqlCompilationManager
import org.opencds.cqf.cql.ls.server.manager.IgContextManager
import org.opencds.cqf.cql.ls.server.manager.LibraryResolutionManager
import org.opencds.cqf.cql.ls.server.plugin.CommandContribution
import org.opencds.cqf.cql.ls.server.utility.Futures
import java.util.concurrent.CompletableFuture

class DebugCommandContribution(
    private val compilationManager: CqlCompilationManager,
    private val contentService: ContentService,
    private val igContextManager: IgContextManager,
    private val libraryResolutionManager: LibraryResolutionManager,
) : CommandContribution {
    companion object {
        const val START_DEBUG_COMMAND = "org.opencds.cqf.cql.debug.startDebugSession"
    }

    private var debugSession: DebugSession? = null

    override fun getCommands(): Set<String> {
        return setOf(START_DEBUG_COMMAND)
    }

    fun stop() {
        debugSession?.stop()
    }

    override fun executeCommand(params: ExecuteCommandParams): CompletableFuture<Any> {
        if (START_DEBUG_COMMAND == params.command) {
            val session = debugSession
            if (session == null || !session.isActive()) {
                val debugServer = CqlDebugServer(compilationManager, contentService, igContextManager, libraryResolutionManager)
                debugServer.enableStreaming()
                val newSession = DebugSession(debugServer)
                debugSession = newSession
                return newSession.start().thenApply { it as Any }
            } else {
                return Futures.failed(
                    IllegalStateException("Please wait for the current debug session to end before starting a new one."),
                )
            }
        } else {
            throw IllegalArgumentException(
                "DebugPlugin doesn't support command ${params.command}",
            )
        }
    }
}
